anomalib/data/custom.py

```diff
@@ -0,0 +1,490 @@
+        extensions (Optional[Tuple[str, ...]], optional): Type of the image extensions to read from the
+            directory.
+
+    Returns:
+        DataFrame: an output dataframe containing samples for the requested split (ie., train or test)
+    """
+    normal_dir = _check_and_convert_path(normal_dir)
+    abnormal_dir = _check_and_convert_path(abnormal_dir)
+
+    if extensions is None:
+        extensions = IMG_EXTENSIONS
+
+    # Get filenames from normal and abnormal directory.
+    normal_filenames = [f for f in normal_dir.glob(r"**/*") if f.suffix in extensions]
+    abnormal_filenames = [f for f in abnormal_dir.glob(r"**/*") if f.suffix in extensions]
+    filenames = normal_filenames + abnormal_filenames
+
+    if len(normal_filenames) == 0:
+        raise RuntimeError(f"Found 0 normal images in {normal_dir}")
+
+    if len(abnormal_filenames) == 0:
+        raise RuntimeError(f"Found 0 annormal images in {abnormal_dir}")
+
+    # Add normal and abnormal labels to the samples as `label` column.
+    normal_labels = ["normal"] * len(normal_filenames)
+    abnormal_labels = ["abnormal"] * len(abnormal_filenames)
+    labels = normal_labels + abnormal_labels
+
+    samples = DataFrame({"image_path": filenames, "label": labels})
+
+    # Create label index for normal (0) and abnormal (1) images.
+    samples.loc[(samples.label == "normal"), "label_index"] = 0
+    samples.loc[(samples.label == "abnormal"), "label_index"] = 1
+    samples.label_index = samples.label_index.astype(int)
+
+    # If a path to mask is provided, add it to the sample dataframe.
+    if mask_dir is not None:
+        mask_dir = _check_and_convert_path(mask_dir)
+        normal_gt = ["" for f in normal_filenames]
+        abnormal_gt = [str(mask_dir / f.name) for f in abnormal_filenames]
+        gt_filenames = normal_gt + abnormal_gt
+
+        samples["mask_path"] = gt_filenames
+
+    # Ensure the pathlib objects are converted to str.
+    # This is because torch dataloader doesn't like pathlib.
+    samples = samples.astype({"image_path": "str"})
+
+    # Create train/test split.
+    # By default, all the normal samples are assigned as train.
+    #   and all the abnormal samples are test.
+    samples.loc[(samples.label == "normal"), "split"] = "train"
+    samples.loc[(samples.label == "abnormal"), "split"] = "test"
+    samples = split_normal_images_in_train_set(
+        samples=samples, split_ratio=split_ratio, seed=seed, normal_label="normal"
+    )
+
+    # If `create_validation_set` is set to True, the test set is split into half.
+    if create_validation_set:
+        samples = create_validation_set_from_test_set(samples, seed=seed, normal_label="normal")
+
+    # Get the data frame for the split.
+    if split is not None and split in ["train", "val", "test"]:
+        samples = samples[samples.split == split]
+        samples = samples.reset_index(drop=True)
+
+    return samples
+
+
+class CustomDataset(Dataset):
```

Comment:
I'm not sure about the naming. Maybe `FolderDataset` would be more appropriate? Custom sounds a bit like users can choose their own 'custom' format. But this class represents a dataset that follows a fixed format based on the folder structure of the data.